src/toxcore/dht/server/mod.rs

```diff
@@ -93,136 +106,264 @@ pub struct Server {
+            last_nodes_req_time: Arc::new(RwLock::new(Instant::now())),
             tcp_onion_sink: None
         }
     }
+
+    /// return peers_cache member variable
+    pub fn get_peers_cache(&self) -> &Arc<RwLock<HashMap<PublicKey, ClientData>>> {
+        &self.peers_cache
+    }
+
+    /// add friend
+    pub fn add_friend(&self, friend: DhtFriend) -> IoFuture<()> {
+        let mut friends = self.friends.write();
+
+        friends.push(friend);
+
+        Box::new(future::ok(()))
+    }
+
+    /// main loop of dht server, call this function every second
+    pub fn dht_main_loop(&self, args: DhtMainLoopArgs) -> IoFuture<()> {
+        self.remove_timedout_clients(Duration::from_secs(args.kill_node_timeout));
+        self.remove_timedout_ping_ids(Duration::from_secs(args.ping_timeout));
+
+        let mut friends = self.friends.write();
+
+        let ping_bootstrap_nodes = self.ping_bootstrap_nodes();
+        let ping_and_get_close_nodes = self.ping_and_get_close_nodes(Duration::from_secs(args.ping_interval));
+        let send_nodes_req_random = self.send_nodes_req_random(Duration::from_secs(args.bad_node_timeout), Duration::from_secs(args.nodes_req_interval));
+        let send_nodes_req_packets = {
+            let nodes_sender = friends.iter_mut()
+                .map(|friend| {
+                    friend.send_nodes_req_packets(self, Duration::from_secs(args.ping_interval), Duration::from_secs(args.nodes_req_interval), Duration::from_secs(args.bad_node_timeout))
+                });
+
+            let nodes_stream = stream::futures_unordered(nodes_sender).then(|_| Ok(()));
+
+            nodes_stream.for_each(|()| Ok(()))
+        };
+
+        let res = ping_bootstrap_nodes.join4(ping_and_get_close_nodes,
+                                             send_nodes_req_random,
+                                             send_nodes_req_packets)
+            .map(|((), (), (), ())| ());
+
+        Box::new(res)
+    }
+
+    fn ping_bootstrap_nodes(&self) -> IoFuture<()> {
+        let mut bootstrap_nodes = Bucket::new(None);
+        mem::swap(&mut bootstrap_nodes, self.bootstrap_nodes.write().deref_mut());
+
+        let mut peers_cache = self.peers_cache.write();
+        let peers_cache = peers_cache.deref_mut();
+
+        let nodes_sender = bootstrap_nodes.nodes.iter()
+            .map(|node| {
+                let client = peers_cache.entry(node.pk).or_insert_with(ClientData::new);
+
+                self.send_nodes_req(*node, self.pk, client)
+            });
+
+        let nodes_stream = stream::futures_unordered(nodes_sender).then(|_| Ok(()));
+
+        Box::new(nodes_stream.for_each(|()| Ok(())))
+    }
+
+    fn ping_and_get_close_nodes(&self, ping_interval: Duration) -> IoFuture<()> {
+        let close_nodes = self.close_nodes.read();
+        let mut peers_cache = self.peers_cache.write();
+
+        let nodes_sender = close_nodes.iter()
+            .map(|node| {
```

Comment:
use `.filter` by `client.last_ping_req_time.elapsed() >= ping_interval` and then `.map` values and send nodes request for each item